### PR TITLE
Support vNic type and Reservation affitinity

### PR DESCRIPTION
**Pull request by gogasca**
Now you can define the desired vNIC type and Reservation affinity features from Compute Engine when creating your Notebook instance.

Support vNic type and Reservation affinity features from Compute Engine into Notebook instance creation request
API [reference](https://cloud.google.com/notebooks/docs/reference/rpc/google.cloud.notebooks.v1#google.cloud.notebooks.v1.Instance)

[vNIC type](https://cloud.google.com/compute/docs/networking/using-gvnic#create-instance-gvnic-image)
[Compute Reservations](https://cloud.google.com/compute/docs/instances/reserving-zonal-resources)

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
`notebooks`: added support for `nic_type`, `reservation_affinity` to `google_notebooks_instance`
```

### CODE REVIEW

File: mmv1/products/notebooks/api.yaml
Comment (melinath) [inline mmv1/products/notebooks/api.yaml:298]:
```suggestion
      name: 'ReservationAffinity'
```

**Review by melinath — CHANGES_REQUESTED** [review]
Hi, @gogasca - thanks for your contribution! Please make sure the new fields you're adding are exercised by tests - by modifying an existing example (defined in [notebooks/terraform.yaml](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/products/notebooks/terraform.yaml)), adding a new example, or modifying the [handwritten test file](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/third_party/terraform/tests/resource_notebooks_instance_test.go.erb)

**Review by melinath — CHANGES_REQUESTED** [review]
In addition to adding tests as described in the previous change request, it looks like your release notes are not properly formatted. Please review the release notes guide & update them. 

It also looks like the indentation in your yaml file is incorrect - please make sure that you're able to run the compiler locally as described in this repository's README.

**Comment by gogasca** [comment]
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2FGoogleCloudPlatform%2Fmagic-modules%2Fpull%2F4981) for more info**.

@googlebot I fixed it.

**Comment by gogasca** [comment]
> In addition to adding tests as described in the previous change request, it looks like your release notes are not properly formatted. Please review the release notes guide & update them.
> 
> It also looks like the indentation in your yaml file is incorrect - please make sure that you're able to run the compiler locally as described in [this repository's README](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers).
> 
> Could you link to the documentation for this feature? I don't see it listed on https://cloud.google.com/notebooks/docs/reference/rest/v1beta1/projects.locations.instances#Instance.

With respect to beta, this feature is not in v1beta1 API, what's the best way to exclude it from beta?

**Comment by gogasca** [comment]
Also quick question, not sure why cla is still set to No.
Thanks!

**Review by melinath — DISMISSED** [review]
Per discussion - this feature will need to be supported in both GA and beta; we don't support GA-only features. Marking as a change request to remove it from my review queue.

**Comment by melinath** [comment]
@googlebot I fixed it.

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:284]: This should be a top-level enum called `nicType`, it isn't nested as far as I can tell.

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:298]:
```suggestion
        name: 'reservationAffinity'
```

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:303]:
```suggestion
            name: 'type'
            required: true
```

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:307]:
```suggestion
```

X_UNSPECIFIED is always a dummy value, coming from proto -> rest translation

File: mmv1/templates/terraform/examples/notebook_instance_full.tf.erb
Comment (rileykarson) [inline mmv1/templates/terraform/examples/notebook_instance_full.tf.erb:34]:
```suggestion
  nic_type = "VIRTIO_NET"
```

**Review by rileykarson — CHANGES_REQUESTED** [review]
Seeing some errors, possible corrections inline:

```
Error: Error creating Instance: googleapi: Error 400: Invalid JSON payload received. Unknown name "NicType" at 'instance': Cannot find field.
Invalid JSON payload received. Unknown name "ReservationAffinity" at 'instance': Cannot find field.
Details:
[
{
"@type": "type.googleapis.com/google.rpc.BadRequest",
"fieldViolations": [
{
"description": "Invalid JSON payload received. Unknown name \"NicType\" at 'instance': Cannot find field.",
"field": "instance"
},
{
"description": "Invalid JSON payload received. Unknown name \"ReservationAffinity\" at 'instance': Cannot find field.",
"field": "instance"
}
]
}
]
on terraform_plugin_test.tf line 2, in resource "google_notebooks_instance" "instance":
2: resource "google_notebooks_instance" "instance" {
--- FAIL: TestAccNotebooksInstance_notebookInstanceFullExample (3.70s)
```

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:288]: Strings can't have subproperties, you'll want to remove these.

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:289]:
```suggestion
```

**Comment by rileykarson** [comment]
Waiting on [VCR] test results.

File: mmv1/products/notebooks/api.yaml
Comment (gogasca) [reply mmv1/products/notebooks/api.yaml:289]: This is a default that is not standard. UNSPECIFIED_NIC_TYPE vs NIC_TYPE_UNSPECIFIED. This is because we use Compute API under the hood and decided to use same default value. https://cloud.google.com/workflows/docs/reference/googleapis/compute/v1/Overview (Search for UNSPECIFIED_NIC_TYPE)

**Comment by gogasca** [comment]
> It looks like `TestAccNotebooksInstance_notebookInstanceFullExample` is currently failing in VCR with:
> 
> ```
> Error: Error creating Instance: Post "https://notebooks.googleapis.com/v1beta1/projects/<project-id>/locations/<location>/instances?alt=json&instanceId=tf-test-notebooks-instancecjyyl6xydx": Requested interaction not found
> ```
> 
> It looks like this is from the REPLAYING run & indicates a missing cassette; the RECORDING run is never triggered. i.e. it's a CI issue.

Sounds good, just replied to few comments now that Github is back.

**Comment by melinath** [comment]
/gcbrun - CI issue is now fixed

**Comment by melinath** [comment]
It looks like this ended up being queued for too long & didn't re-run the specific tests.

Here's a non-VCR build with just the full example test: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/200665

**Comment by gogasca** [comment]
> It looks like this ended up being queued for too long & didn't re-run the specific tests.
> 
> Here's a non-VCR build with just the full example test: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/200665

Not sure what I need to do here?

**Comment by rileykarson** [comment]
No action for you @gogasca

It turns out a typo in CI config made the test get skipped, I fixed that for the build and kicked it back off. Results will take an hour or two.

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:297]: I can't comment directly on L298 for some reason. Weird. Are you sure the field is called `type`? While the type of the field is called `Type`, the field itself is called `consume_reservation_type` in the RPC docs: https://cloud.google.com/notebooks/docs/reference/rpc/google.cloud.notebooks.v1#google.cloud.notebooks.v1.ReservationAffinity

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:285]: Given that this feature will be in public preview please add `min_version: beta` to each new field, and register a beta version at the top of this file. It should point to the REST API that this feature will be exposed under.

**Review by rileykarson — COMMENTED** [review]
Well, we get an error back if the test fails. Resolution inline.

```
------- Stdout: -------
=== RUN   TestAccNotebooksInstance_notebookInstanceFullExample
=== PAUSE TestAccNotebooksInstance_notebookInstanceFullExample
=== CONT  TestAccNotebooksInstance_notebookInstanceFullExample
provider_test.go:276: Step 1/2 error: Error running apply: exit status 1
2021/08/11 17:08:45 [DEBUG] Using modified User-Agent: Terraform/0.12.30 HashiCorp-terraform-exec/0.13.0
Error: Error creating Instance: googleapi: Error 400: Invalid JSON payload received. Unknown name "type" at 'instance.reservation_affinity': Cannot find field.
Details:
[
{
"@type": "type.googleapis.com/google.rpc.BadRequest",
"fieldViolations": [
{
"description": "Invalid JSON payload received. Unknown name \"type\" at 'instance.reservation_affinity': Cannot find field.",
"field": "instance.reservation_affinity"
}
]
}
]
on terraform_plugin_test.tf line 2, in resource "google_notebooks_instance" "instance":
2: resource "google_notebooks_instance" "instance" {
--- FAIL: TestAccNotebooksInstance_notebookInstanceFullExample (2.70s)
FAIL
```

File: mmv1/products/notebooks/api.yaml
Comment (gogasca) [reply mmv1/products/notebooks/api.yaml:297]: Good catch, changing this now.
Github has been having all sorts of issues since yesterday.

File: mmv1/products/notebooks/api.yaml
Comment (gogasca) [reply mmv1/products/notebooks/api.yaml:285]: ack

**Comment by rileykarson** [comment]
The API is accepting the `nic_type` and `reservation_affinity` values, but is not returning them. Moving forward is blocked until they are returned.

```
------- Stdout: -------
=== RUN   TestAccNotebooksInstance_notebookInstanceFullExample
=== PAUSE TestAccNotebooksInstance_notebookInstanceFullExample
=== CONT  TestAccNotebooksInstance_notebookInstanceFullExample
provider_test.go:276: Step 1/2 error: After applying this test step, the plan was not empty.
stdout:
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement
Terraform will perform the following actions:
# google_notebooks_instance.instance must be replaced
-/+ resource "google_notebooks_instance" "instance" {
boot_disk_size_gb  = 110
boot_disk_type     = "PD_SSD"
~ create_time        = "2021-08-11T18:37:20.571975308Z" -> (known after apply)
~ id                 = "projects/ci-test-project-188019/locations/us-central1-a/instances/tf-test-notebooks-instance9stbsw1iss" -> (known after apply)
install_gpu_driver = true
instance_owners    = [
"admin@hashicorptest.com",
]
~ labels             = {
- "goog-caip-notebook" = "" -> null
"k"                  = "val"
}
location           = "us-central1-a"
machine_type       = "e2-medium"
metadata           = {
"terraform" = "true"
}
name               = "tf-test-notebooks-instance9stbsw1iss"
~ network            = "https://www.googleapis.com/compute/v1/projects/ci-test-project-188019/global/networks/default" -> "projects/ci-test-project-188019/global/networks/default"
+ nic_type           = "VIRTIO_NET"
no_proxy_access    = true
no_public_ip       = true
~ project            = "ci-test-project-188019" -> (known after apply)
+ proxy_uri          = (known after apply)
service_account    = "hashicorp-test-runner@ci-test-project-188019.iam.gserviceaccount.com"
~ state              = "PROVISIONING" -> (known after apply)
~ subnet             = "https://www.googleapis.com/compute/v1/projects/ci-test-project-188019/regions/us-central1/subnetworks/default" -> "projects/ci-test-project-188019/regions/us-central1/subnetworks/default"
~ update_time        = "2021-08-11T18:37:44.138406509Z" -> (known after apply)
+ reservation_affinity {
+ consume_reservation_type = "NO_RESERVATION"
}
+ shielded_instance_config {
+ enable_integrity_monitoring = (known after apply)
+ enable_secure_boot          = (known after apply)
+ enable_vtpm                 = (known after apply)
}
~ vm_image {
image_family = "tf-latest-cpu"
project      = "deeplearning-platform-release"
}
}
Plan: 1 to add, 0 to change, 1 to destroy.
--- FAIL: TestAccNotebooksInstance_notebookInstanceFullExample (67.67s)
FAIL
```

**Comment by gogasca** [comment]
> The API is accepting the `nic_type` and `reservation_affinity` values, but is not returning them. Moving forward is blocked until they are returned.
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccNotebooksInstance_notebookInstanceFullExample
> === PAUSE TestAccNotebooksInstance_notebookInstanceFullExample
> === CONT  TestAccNotebooksInstance_notebookInstanceFullExample
> provider_test.go:276: Step 1/2 error: After applying this test step, the plan was not empty.
> stdout:
> An execution plan has been generated and is shown below.
> Resource actions are indicated with the following symbols:
> -/+ destroy and then create replacement
> Terraform will perform the following actions:
> # google_notebooks_instance.instance must be replaced
> -/+ resource "google_notebooks_instance" "instance" {
> boot_disk_size_gb  = 110
> boot_disk_type     = "PD_SSD"
> ~ create_time        = "2021-08-11T18:37:20.571975308Z" -> (known after apply)
> ~ id                 = "projects/ci-test-project-188019/locations/us-central1-a/instances/tf-test-notebooks-instance9stbsw1iss" -> (known after apply)
> install_gpu_driver = true
> instance_owners    = [
> "admin@hashicorptest.com",
> ]
> ~ labels             = {
> - "goog-caip-notebook" = "" -> null
> "k"                  = "val"
> }
> location           = "us-central1-a"
> machine_type       = "e2-medium"
> metadata           = {
> "terraform" = "true"
> }
> name               = "tf-test-notebooks-instance9stbsw1iss"
> ~ network            = "https://www.googleapis.com/compute/v1/projects/ci-test-project-188019/global/networks/default" -> "projects/ci-test-project-188019/global/networks/default"
> + nic_type           = "VIRTIO_NET"
> no_proxy_access    = true
> no_public_ip       = true
> ~ project            = "ci-test-project-188019" -> (known after apply)
> + proxy_uri          = (known after apply)
> service_account    = "hashicorp-test-runner@ci-test-project-188019.iam.gserviceaccount.com"
> ~ state              = "PROVISIONING" -> (known after apply)
> ~ subnet             = "https://www.googleapis.com/compute/v1/projects/ci-test-project-188019/regions/us-central1/subnetworks/default" -> "projects/ci-test-project-188019/regions/us-central1/subnetworks/default"
> ~ update_time        = "2021-08-11T18:37:44.138406509Z" -> (known after apply)
> + reservation_affinity {
> + consume_reservation_type = "NO_RESERVATION"
> }
> + shielded_instance_config {
> + enable_integrity_monitoring = (known after apply)
> + enable_secure_boot          = (known after apply)
> + enable_vtpm                 = (known after apply)
> }
> ~ vm_image {
> image_family = "tf-latest-cpu"
> project      = "deeplearning-platform-release"
> }
> }
> Plan: 1 to add, 0 to change, 1 to destroy.
> --- FAIL: TestAccNotebooksInstance_notebookInstanceFullExample (67.67s)
> FAIL
> ```

Ack, adding this info in API. Will rollout ASAP and update

**Comment by gogasca** [comment]
API returns now correct information:

```
{
  "name": "projects/gogasca-autopush-notebooks/locations/us-west1-a/operations/operation-1629325528306-5c9dce956950e-0d51ed84-3d4f87b5",
  "metadata": {
    "@type": "type.googleapis.com/google.cloud.notebooks.v1beta1.OperationMetadata",
    "createTime": "2021-08-18T22:25:28.528456286Z",
    "endTime": "2021-08-18T22:25:50.302220904Z",
    "target": "projects/gogasca-autopush-notebooks/locations/us-west1-a/instances/ref-3a",
    "verb": "create",
    "requestedCancellation": false,
    "apiVersion": "v1beta1",
    "endpoint": "CreateInstance"
  },
  "done": true,
  "response": {
    "@type": "type.googleapis.com/google.cloud.notebooks.v1beta1.Instance",
    "name": "projects/gogasca-autopush-notebooks/locations/us-west1-a/instances/ref-3a",
    "serviceAccount": "634850243685-compute@developer.gserviceaccount.com",
    "machineType": "https://www.googleapis.com/compute/v1/projects/gogasca-autopush-notebooks/zones/us-west1-a/machineTypes/n1-standard-1",
    "state": "PROVISIONING",
    "network": "https://www.googleapis.com/compute/v1/projects/gogasca-autopush-notebooks/global/networks/default",
    "subnet": "https://www.googleapis.com/compute/v1/projects/gogasca-autopush-notebooks/regions/us-west1/subnetworks/default",
    "labels": {
      "goog-caip-notebook": ""
    },
    "metadata": {
      "shutdown-script": "/opt/deeplearning/bin/shutdown_script.sh",
      "notebooks-api": "AUTOPUSH",
      "proxy-mode": "service_account",
      "proxy-user-mail": "gogasca@google.com",
      "enable-guest-attributes": "TRUE",
      "report-system-health": "TRUE"
    },
    "createTime": "2021-08-18T22:25:28.523521866Z",
    "updateTime": "2021-08-18T22:25:28.523521866Z",
    "nicType": "VIRTIO_NET",
    "reservationAffinity": {
      "consumeReservationType": "NO_RESERVATION"
    }
  }
}
```

**Comment by gogasca** [comment]
Checking if this time we have better luck?

**Comment by rileykarson** [comment]
Ah, looks like generation failed- can you rebase this PR? That should fix it.

**Comment by gogasca** [comment]
Made a small change.

**Comment by rileykarson** [comment]
I specifically need you to rebase so that the build definitions under the `.ci/` folder are updated.

**Comment by gogasca** [comment]
> I specifically need you to rebase so that the build definitions under the `.ci/` folder are updated.

Done

**Comment by rileykarson** [comment]
Man, CI is not happy w/ me right now. Gonna try again.

/gcbrun

**Comment by rileykarson** [comment]
Over-aggressive `Fixes: ` message, GitHub didn't indicate it would do that!

**Comment by ScottSuarez** [comment]
> Over-aggressive Fixes: message, GitHub didn't indicate it would do that!

yeah i don't know why they got rid of the squiggly thing

**Comment by ScottSuarez** [comment]
oh wait, an closing an issue can close prs.... weird

**Comment by gogasca** [comment]
I assume at this point I just need to wait for another test?

File: mmv1/products/notebooks/api.yaml
Comment (rileykarson) [inline mmv1/products/notebooks/api.yaml:299]:
```suggestion
            name: 'consumeReservationType'
```

This caused us to send the message

```
 "reservationAffinity": {
  "consume_reservation_type": "NO_RESERVATION"
 },
```

instead of

```
 "reservationAffinity": {
  "consumeReservationType": "NO_RESERVATION"
 },
```

and then look for the key `consume_reservation_type` instead of `consumeReservationType` in the response.

**Review by rileykarson — APPROVED** [review]

**Comment by rileykarson** [comment]
@melinath can you dismiss your change request? Those are set to block merges.

**Comment by melinath** [comment]
@rileykarson done

**Pull request merged**
